File: solutions/elixir/anagram/2/anagram.exs
defmodule Anagram do
  def match(word, list) do
    Enum.filter(list, is_anagram(word, &1))
  end

  defp is_anagram(word, candidate_word) do
    has_same_letters(word, candidate_word) && !is_same_word(word, candidate_word)
  end

  defp has_same_letters(word, candidate_word) do
    word_letters(word) == word_letters(candidate_word)
  end
  defp word_letters(word) do
    :lists.sort String.codepoints word |> String.downcase
  end

  defp is_same_word(word, candidate_word) do
    String.downcase(word) == String.downcase(candidate_word)
  end
end

File: solutions/javascript/anagram/2/anagram.js
"use strict";
var Anagram = (function() {
  function Anagram(word) {
    var self = this;
    this.word = word;
    this.match = function(words) {
      var matches = words.reduce(function(collection, candidateWord) {
        if (isAnagram(candidateWord)) { collection.push(candidateWord); }
        return collection;
      }, []);
      return matches;
    }
    function isAnagram(candidateWord) {
      return isSameLetters(candidateWord) && !isSameWord(candidateWord);
    }
    function isSameLetters(candidateWord) {
      return reorderLetters(candidateWord) === reorderLetters(self.word);
    }
    function isSameWord(candidateWord) {
      return candidateWord.toLowerCase() === self.word.toLowerCase();
    }
    function reorderLetters(word) {
      return word.toLowerCase().split("").sort().join();
    }
  }
  return Anagram;
})();

module.exports = Anagram;